### ISSUE TITLE
cimi external objects should have a content-type attribute

### ISSUE THREAD

**Issue by 0xbase12** [slipstream/SlipStreamServer#1409]
actually the filename is external object id

**Comment by 0xbase12** [comment]
Cal suggestion:
For the file type, I think that it would be better to add an optional mimeType field to the ExternalObject and let the user set this when creating the resource.  With S3 it is possible to set metadata on the object, so I think that when we create the object this mimeType could be provided.  In this case, S3 should provide the right mimeType on download.

**Comment by loomis** [comment]
Looking at the S3 features (at least for AWS), every object has associated metadata.  This is a set of key/value pairs, so setting "Content-Type" and then the value of the metadata should work correctly.  I haven't looked at how this would work with the Amazonica API.

**Comment by elegoff** [comment]
I tried the following :

**Comment by elegoff** [comment]
FWIW sharing a basic test :

```
;; Upload Json content to S3
(clj-http.client/put (generate-url "slipstream-bucket-name" "akey" 10 true) 
{:body (json/edn->json {:foo :bar}) 
:content-type :json})
```

;;Fetch content-type for same  doc
```
(-> (clj-http.client/get (generate-url "slipstream-bucket-name" "akey" 10 false)) 
    :headers 
    (get "Content-Type"))
=> "application/json"
```

**Comment by loomis** [comment]
The optional attribute name should be `contentType` and be limited to valid values for the `Content-Type` HTTP header.  (Probably non-blank string because values like `text/html; charset=utf-8` are permitted. 

The content type must be added to the generated URL itself, because we cannot otherwise guarantee that the content-type value will be that specified in the external object resource.  This can be done with the [Java API](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/GeneratePresignedUrlRequest.html) using the `setContentType` method.  The value must not be added as a user-level key/value pair.

**Comment by elegoff** [comment]
Investigations regarding the addition of `Content-Type` led to the same conclusion 
either via Amazonica or by using AWS Java SDK
(Actually adding the `:content-type` key in `s3/generate-presigned-url` has same effect as using the `setContentType` [ java API](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/GeneratePresignedUrlRequest.html#setContentType-java.lang.String-) )

When creating pre-signed URLs, the headers that are present when the request is signed 
MUST also be present when using the pre-signed URL. 

As in the [source comment](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/GeneratePresignedUrlRequest.java#L489)
 : 
> "The content-type is included in the signature."

This means that creating things like pre-signed PUT requests 
with a content-type require 
that the both the  HTTP verb (PUT) and the  same content-type MUST be used.

 
Otherwise error 403 is raised such as ```<Error><Code>SignatureDoesNotMatch</Code>```

Illustration with code :

```
;; Content-type specific PUT URL
(def upload-url (str (s3/generate-presigned-url (get-s3-cred)
                                {:bucket-name "slipstream-bucket-name"
                                 :key "test07031008"
                                 :method "PUT"
                                 :content-type "image/jpg"
                                 }
                                (-> 20 time/minutes time/from-now))))

;;Setting no content-type fails
(def no-content-type-resp (clj-http.client/put upload-url))
                                =>CompilerException clojure.lang.ExceptionInfo: clj-http: status 403 {:request-time 273, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? true, :type :clj-http.client/unexceptional-status, :reason-phrase "Forbidden", :headers {"Server" "nginx", "Date" "Wed, 07 Mar 2018 13:59:41 GMT", "Content-Type" "application/xml", "Transfer-Encoding" "chunked", "Connection" "close", "x-amz-bucket-region" "ch-dk-2", "x-amz-request-id" "845e758d-ee6b-4cf7-85de-3c515bbbac00", "x-amzn-requestid" "845e758d-ee6b-4cf7-85de-3c515bbbac00", "x-amz-id-2" "845e758d-ee6b-4cf7-85de-3c515bbbac00"}, :orig-content-encoding nil, :status 403, :length -1, :body "<?xml version=\"1.0\" encoding=\"UTF-8\"?><Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><StringToSign>AWS4-HMAC-SHA256\n20180307T134852Z\n20180307/us-east-1/s3/aws4_request\n7437ce8aae260089cae053aaf7b03e6a0492e5a418c9bfbb83851d6f518aca35</StringToSign><StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 31 38 30 33 30 37 54 31 33 34 38 35 32 5a 0a 32 30 31 38 30 33 30 37 2f 75 73 2d 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 37 34 33 37 63 65 38 61 61 65 32 36 30 30 38 39 63 61 65 30 35 33 61 61 66 37 62 30 33 65 36 61 30 34 39 32 65 35 61 34 31 38 63 39 62 66 62 62 38 33 38 35 31 64 36 66 35 31 38 61 63 61 33 35</StringToSignBytes><RequestId>845e758d-ee6b-4cf7-85de-3c515bbbac00</RequestId><HostId>845e758d-ee6b-4cf7-85de-3c515bbbac00</HostId></Error>", :trace-redirects []}, compiling:(form-init938337887379824303.clj:3:3) 
                                
;;Using a Content-Type different from what was included in URL signature fails 
(def wrong-ct (clj-http.client/put upload-url {:content-type "text/html; charset=utf-8"}))
                                => CompilerException clojure.lang.ExceptionInfo: clj-http: status 403 {:request-time 280, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :chunked? true, :type :clj-http.client/unexceptional-status, :reason-phrase "Forbidden", :headers {"Server" "nginx", "Date" "Wed, 07 Mar 2018 13:55:56 GMT", "Content-Type" "application/xml", "Transfer-Encoding" "chunked", "Connection" "close", "x-amz-bucket-region" "ch-dk-2", "x-amz-request-id" "922b152c-211c-49f0-ae60-3897159fbeee", "x-amzn-requestid" "922b152c-211c-49f0-ae60-3897159fbeee", "x-amz-id-2" "922b152c-211c-49f0-ae60-3897159fbeee"}, :orig-content-encoding nil, :status 403, :length -1, :body "<?xml version=\"1.0\" encoding=\"UTF-8\"?><Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><StringToSign>AWS4-HMAC-SHA256\n20180307T134852Z\n20180307/us-east-1/s3/aws4_request\n89acf37bb133022525eda1235753e7b38921ddcbb1dd2fec13b76e105d96edd2</StringToSign><StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 31 38 30 33 30 37 54 31 33 34 38 35 32 5a 0a 32 30 31 38 30 33 30 37 2f 75 73 2d 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 38 39 61 63 66 33 37 62 62 31 33 33 30 32 32 35 32 35 65 64 61 31 32 33 35 37 35 33 65 37 62 33 38 39 32 31 64 64 63 62 62 31 64 64 32 66 65 63 31 33 62 37 36 65 31 30 35 64 39 36 65 64 64 32</StringToSignBytes><RequestId>922b152c-211c-49f0-ae60-3897159fbeee</RequestId><HostId>922b152c-211c-49f0-ae60-3897159fbeee</HostId></Error>", :trace-redirects []}, compiling:(form-init938337887379824303.clj:1:17)                                 

;;Setting expected content-type succeeds
    (clj-http.client/put upload-url {:content-type "image/jpg"}))
                               =>
                                {:request-time 409,
                                 :repeatable? false,
                                 :protocol-version {:name "HTTP", :major 1, :minor 1},
                                 :streaming? false,
                                 :chunked? false,
                                 :reason-phrase "OK",
                                 :headers {"Server" "nginx",
                                           "Date" "Wed, 07 Mar 2018 14:05:25 GMT",
                                           "Content-Length" "0",
                                           "Connection" "close",
                                           "x-amz-bucket-region" "ch-dk-2",
                                           "etag" "\"d41d8cd98f00b204e9800998ecf8427e\"",
                                           "x-amz-request-id" "d6a780bf-efbc-4338-987e-3c25a80a5400",
                                           "x-amz-requestid" "d6a780bf-efbc-4338-987e-3c25a80a5400",
                                           "x-amz-id-2" "d6a780bf-efbc-4338-987e-3c25a80a5400"},
                                 :orig-content-encoding nil,
                                 :status 200,
                                 :length 0,
                                 :body "",
                                 :trace-redirects []}
```

In other words , it would seem the Pre Signed URLs design for S3 goes against the requirement 

> "The content-type value must not be added as a user-level key/value pair."

**Comment by loomis** [comment]
> In other words , it would seem the Pre Signed URLs design for S3 goes against the requirement

>> "The content-type value must not be added as a user-level key/value pair."

This refers to the metadata keys that start with "x-amz-meta-".  It is a shame that the content-type isn't included directly in the URL, but if documented, the user having to add the correct content-type header isn't too much of a burden.
